File: shine-server-db/src/main/java/shine/db/channels/ChannelNameRules.java
package shine.db.channels;

import java.util.Locale;
import java.util.regex.Pattern;

public final class ChannelNameRules {
    private static final int MIN_DISPLAY_NAME_LENGTH = 3;
    private static final int MAX_DISPLAY_NAME_LENGTH = 32;
    private static final Pattern DISPLAY_ALLOWED_PATTERN =
            Pattern.compile("^[\\p{IsLatin}\\p{IsCyrillic}0-9 _-]+$");

    private ChannelNameRules() {}

    public static String normalizeDisplayName(String value) {
        if (value == null) return "";
        return value.trim().replaceAll("\\s+", " ");
    }

    public static String requireValidDisplayNameForCreate(String rawName) {
        String normalized = normalizeDisplayName(rawName);
        if (normalized.isEmpty()) {
            throw new IllegalArgumentException("channelName is blank");
        }

        int length = normalized.codePointCount(0, normalized.length());
        if (length < MIN_DISPLAY_NAME_LENGTH || length > MAX_DISPLAY_NAME_LENGTH) {
            throw new IllegalArgumentException("channelName length must be 3..32");
        }

        if (!DISPLAY_ALLOWED_PATTERN.matcher(normalized).matches()) {
            throw new IllegalArgumentException("channelName contains unsupported characters");
        }

        return normalized;
    }

    public static String toCanonicalSlug(String rawName) {
        String normalized = normalizeDisplayName(rawName);
        if (normalized.isEmpty()) {
            throw new IllegalArgumentException("channelName is blank");
        }

        String lowered = normalized.toLowerCase(Locale.ROOT).replace('\u0451', '\u0435');
        StringBuilder slug = new StringBuilder(lowered.length());
        boolean pendingSeparator = false;

        for (int i = 0; i < lowered.length(); ) {
            int cp = lowered.codePointAt(i);
            i += Character.charCount(cp);

            if (cp == ' ' || cp == '_' || cp == '-') {
                pendingSeparator = slug.length() > 0;
                continue;
            }

            if (!isLatinOrCyrillicOrDigit(cp)) {
                throw new IllegalArgumentException("channelName contains unsupported characters");
            }

            if (pendingSeparator && slug.length() > 0) {
                slug.append('-');
            }
            pendingSeparator = false;
            slug.appendCodePoint(cp);
        }

        int len = slug.length();
        if (len > 0 && slug.charAt(len - 1) == '-') {
            slug.deleteCharAt(len - 1);
        }

        if (slug.length() == 0) {
            throw new IllegalArgumentException("channelName canonical slug is empty");
        }

        return slug.toString();
    }

    private static boolean isLatinOrCyrillicOrDigit(int cp) {
        if (Character.isDigit(cp)) return true;
        Character.UnicodeScript script = Character.UnicodeScript.of(cp);
        return script == Character.UnicodeScript.LATIN || script == Character.UnicodeScript.CYRILLIC;
    }
}

File: shine-UI/js/services/channel-name-rules.js
const MIN_LEN = 3;
const MAX_LEN = 32;
const ALLOWED_CHARS_RE = /^[\p{Script=Latin}\p{Script=Cyrillic}0-9 _-]+$/u;

export function normalizeChannelDisplayName(value) {
  if (value == null) return '';
  return String(value).trim().replace(/\s+/g, ' ');
}

export function normalizeChannelDescription(value) {
  if (value == null) return '';
  return String(value).trim().replace(/\s+/g, ' ');
}

export function toCanonicalChannelSlug(value) {
  const normalized = normalizeChannelDisplayName(value);
  if (!normalized) return '';

  const lowered = normalized.toLowerCase().replace(/\u0451/g, '\u0435');
  let out = '';
  let pendingSeparator = false;

  for (const ch of lowered) {
    if (ch === ' ' || ch === '_' || ch === '-') {
      pendingSeparator = out.length > 0;
      continue;
    }
    if (!/[\p{Script=Latin}\p{Script=Cyrillic}0-9]/u.test(ch)) {
      return '';
    }
    if (pendingSeparator && out.length > 0) out += '-';
    out += ch;
    pendingSeparator = false;
  }

  return out.replace(/-+$/g, '');
}

export function validateChannelDisplayName(value) {
  const normalized = normalizeChannelDisplayName(value);
  if (!normalized) {
    return { ok: false, code: 'blank', normalized: '', slug: '' };
  }

  const length = Array.from(normalized).length;
  if (length < MIN_LEN) {
    return { ok: false, code: 'too_short', normalized, slug: '' };
  }
  if (length > MAX_LEN) {
    return { ok: false, code: 'too_long', normalized, slug: '' };
  }
  if (!ALLOWED_CHARS_RE.test(normalized)) {
    return { ok: false, code: 'bad_chars', normalized, slug: '' };
  }
  if (normalized === '0') {
    return { ok: false, code: 'reserved', normalized, slug: '' };
  }

  const slug = toCanonicalChannelSlug(normalized);
  if (!slug) {
    return { ok: false, code: 'bad_chars', normalized, slug: '' };
  }

  return { ok: true, code: '', normalized, slug };
}

export function channelNameErrorText(code) {
  switch (String(code || '').trim()) {
    case 'blank':
      return 'Введите название канала.';
    case 'too_short':
      return 'Название слишком короткое: минимум 3 символа.';
    case 'too_long':
      return 'Название слишком длинное: максимум 32 символа.';
    case 'bad_chars':
      return 'Разрешены кириллица, латиница, цифры, пробел, _ и -.';
    case 'reserved':
      return 'Название "0" зарезервировано.';
    default:
      return 'Некорректное название канала.';
  }
}

export function channelDescriptionErrorText(value) {
  const normalized = normalizeChannelDescription(value);
  if (new TextEncoder().encode(normalized).length > 200) {
    return 'Описание слишком длинное: максимум 200 байт UTF-8.';
  }
  return '';
}
